soci/ztoc_builder.go

```diff
@@ -81,34 +83,113 @@ func BuildZtoc(gzipFile string, span int64, cfg *buildConfig) (*Ztoc, error) {
+
+	metadataOffsetList := make([]flatbuffers.UOffsetT, len(ztoc.Metadata))
+	for i := len(ztoc.Metadata) - 1; i >= 0; i-- {
+		me := ztoc.Metadata[i]
+		name := builder.CreateString(me.Name)
+		t := builder.CreateString(me.Type)
+		linkName := builder.CreateString(me.Linkname)
+		uname := builder.CreateString(me.Uname)
+		gname := builder.CreateString(me.Gname)
+		modTimeBinary, _ := me.ModTime.MarshalText()
+		modTime := builder.CreateString(string(modTimeBinary))
+
+		// preparing Xattrs
+		keys := make([]string, 0, len(me.Xattrs))
+		for k := range me.Xattrs {
+			keys = append(keys, k)
+		}
+		sort.Strings(keys)
+
+		xattrOffsetList := make([]flatbuffers.UOffsetT, 0, len(me.Xattrs))
+		for _, key := range keys {
+			keyOffset := builder.CreateString(key)
+			valueOffset := builder.CreateString(me.Xattrs[key])
+			ztoc_flatbuffers.XattrStart(builder)
+			ztoc_flatbuffers.XattrAddKey(builder, keyOffset)
+			ztoc_flatbuffers.XattrAddValue(builder, valueOffset)
+			xattrOffset := ztoc_flatbuffers.XattrEnd(builder)
+			xattrOffsetList = append(xattrOffsetList, xattrOffset)
+		}
+		ztoc_flatbuffers.FileMetadataStartXattrsVector(builder, len(xattrOffsetList))
+		for j := len(xattrOffsetList) - 1; j >= 0; j-- {
+			builder.PrependUOffsetT(xattrOffsetList[j])
+		}
+		xattrs := builder.EndVector(len(me.Xattrs))
+
+		// preparing the individual file medatada element
+		ztoc_flatbuffers.FileMetadataStart(builder)
+		ztoc_flatbuffers.FileMetadataAddName(builder, name)
+		ztoc_flatbuffers.FileMetadataAddType(builder, t)
+		ztoc_flatbuffers.FileMetadataAddUncompressedOffset(builder, int64(me.UncompressedOffset))
+		ztoc_flatbuffers.FileMetadataAddUncompressedSize(builder, int64(me.UncompressedSize))
+		ztoc_flatbuffers.FileMetadataAddSpanStart(builder, int32(me.SpanStart))
+		ztoc_flatbuffers.FileMetadataAddSpanEnd(builder, int32(me.SpanEnd))
+		ztoc_flatbuffers.FileMetadataAddFirstSpanHasBits(builder, me.FirstSpanHasBits)
+		ztoc_flatbuffers.FileMetadataAddLinkname(builder, linkName)
+		ztoc_flatbuffers.FileMetadataAddMode(builder, me.Mode)
+		ztoc_flatbuffers.FileMetadataAddUid(builder, int32(me.UID))
+		ztoc_flatbuffers.FileMetadataAddGid(builder, int32(me.GID))
+		ztoc_flatbuffers.FileMetadataAddUname(builder, uname)
+		ztoc_flatbuffers.FileMetadataAddGname(builder, gname)
+		ztoc_flatbuffers.FileMetadataAddModTime(builder, modTime)
+		ztoc_flatbuffers.FileMetadataAddDevmajor(builder, me.Devmajor)
+		ztoc_flatbuffers.FileMetadataAddDevminor(builder, me.Devminor)
+		// also add xattrs
+		ztoc_flatbuffers.FileMetadataAddXattrs(builder, xattrs)
+
+		off := ztoc_flatbuffers.FileMetadataEnd(builder)
+		metadataOffsetList[i] = off
 	}
-
-	if _, err := zs.Write(serializedBuf.Bytes()); err != nil {
-		return nil, ocispec.Descriptor{}, fmt.Errorf("cannot compress ztoc: %w", err)
+	ztoc_flatbuffers.ZtocStartMetadataVector(builder, len(ztoc.Metadata))
+	for i := len(metadataOffsetList) - 1; i >= 0; i-- {
+		builder.PrependUOffsetT(metadataOffsetList[i])
```

Comment:
any reason why this cannot be in the loop above?

Reply:
For flatbuffers, you first need to prepare the list of the offsets, then proceed with the placement. You cannot start vector and then start preparing offsets and placing them.

Reply:
:+1: